Syntactically incorrect code suggestion for missing semicolon in `if` without `else`

### Code

```Rust
struct Page;
impl Page {
    fn method(&self) -> usize { 7 }
}
pub fn f(a: i32, page: &Page) {
    if a > 0 {
        page.method()  // oops, missing semicolon
    }
    println!("done");
}
```

### Current output

```Shell
Compiling playground v0.0.1 (/playground)
error[E0308]: mismatched types
 --> src/lib.rs:7:9
  |
6 | /     if a > 0 {
7 | |         page.method()  // oops, missing semicolon
  | |         ^^^^^^^^^^^^^ expected `()`, found `usize`
8 | |     }
  | |_____- `if` expressions without `else` arms expect their inner expression to be `()`
  |
help: consider using a semicolon here
  |
7 |         page.method();  // oops, missing semicolon
  |                      +
help: consider using a semicolon here
  |
8 |     };
  |      +

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` (lib) due to 1 previous error
```

### Desired output

```Shell
Compiling playground v0.0.1 (/playground)
error[E0308]: mismatched types
 --> src/lib.rs:7:9
  |
6 | /     if a > 0 {
7 | |         page.method()  // oops, missing semicolon
  | |         ^^^^^^^^^^^^^ expected `()`, found `usize`
8 | |     }
  | |_____- `if` expressions without `else` arms expect their inner expression to be `()`
  |
help: consider using a semicolon here
  |
7 |         page.method();  // oops, missing semicolon
  |                      +

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` (lib) due to 1 previous error
```

### Rationale and extra context

The diagnostic suggests two possible places to add a semicolon.

But the second one does not work. Adding a semicolon there doesn't fix the problem.

(It's also less likely to be what the user intended than the first suggestion.)

### Rust Version

```Shell
$ rustc +nightly --version --verbose
rustc 1.97.0-nightly (7c3c88f42 2026-05-14)
binary: rustc
commit-hash: 7c3c88f42ad444f4688b865591d84660be4ece2f
commit-date: 2026-05-14
host: x86_64-unknown-linux-gnu
release: 1.97.0-nightly
LLVM version: 22.1.4
```


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_hir_typeck/src/coercion.rs
+++ b/compiler/rustc_hir_typeck/src/coercion.rs
@@ -2051,7 +2051,10 @@ impl<'tcx> CoerceMany<'tcx> {
                     err.span_label(cond_expr.span, "expected this to be `()`");
                 }
                 if expr.can_have_side_effects() {
-                    fcx.suggest_semicolon_at_end(cond_expr.span, &mut err);
+                    // Don't suggest semicolon after if expressions as it does not fix the issue
+                    if !matches!(cond_expr.kind, hir::ExprKind::If(..)) {
+                        fcx.suggest_semicolon_at_end(cond_expr.span, &mut err);
+                    }
                 }
             }
         }

--- a/tests/ui/parser/expr-as-stmt-2.stderr
+++ b/tests/ui/parser/expr-as-stmt-2.stderr
@@ -136,10 +136,6 @@ help: parentheses are required to parse this as an expression
    |
 LL |     (if true { bar() } else { bar() }) && if true { bar() } else { bar() };
    |     +                                +
-help: consider using a semicolon here
-   |
-LL |     if true { bar() } else { bar() }; && if true { bar() } else { bar() };
-   |                                     +
 
 error[E0308]: mismatched types
   --> $DIR/expr-as-stmt-2.rs:24:30
@@ -154,46 +150,26 @@ help: parentheses are required to parse this as an expression
    |
 LL |     (if true { bar() } else { bar() }) && if true { bar() } else { bar() };
    |     +                                +
-help: consider using a semicolon here
-   |
-LL |     if true { bar() } else { bar() }; && if true { bar() } else { bar() };
-   |                                     +
 
 error[E0308]: mismatched types
   --> $DIR/expr-as-stmt-2.rs:27:15
    |
 LL |     if true { bar() } else { bar() } if true { bar() } else { bar() };
    |     ----------^^^^^-----------------
-   |     |         |
+   |     |         |    |
+   |     |         |    help: consider using a semicolon here: `;`
    |     |         expected `()`, found `bool`
    |     expected this to be `()`
-   |
-help: consider using a semicolon here
-   |
-LL |     if true { bar(); } else { bar() } if true { bar() } else { bar() };
-   |                    +
-help: consider using a semicolon here
-   |
-LL |     if true { bar() } else { bar() }; if true { bar() } else { bar() };
-   |                                     +
 
 error[E0308]: mismatched types
   --> $DIR/expr-as-stmt-2.rs:27:30
    |
 LL |     if true { bar() } else { bar() } if true { bar() } else { bar() };
    |     -------------------------^^^^^--
-   |     |                        |
+   |     |                        |    |
+   |     |                        |    help: consider using a semicolon here: `;`
    |     |                        expected `()`, found `bool`
    |     expected this to be `()`
-   |
-help: consider using a semicolon here
-   |
-LL |     if true { bar() } else { bar(); } if true { bar() } else { bar() };
-   |                                   +
-help: consider using a semicolon here
-   |
-LL |     if true { bar() } else { bar() }; if true { bar() } else { bar() };
-   |                                     +
 
 error: aborting due to 13 previous errors
 

--- a/tests/ui/return/issue-82612-return-mutable-reference.stderr
+++ b/tests/ui/return/issue-82612-return-mutable-reference.stderr
@@ -14,10 +14,6 @@ help: consider using a semicolon here
    |
 LL |             value.get_or_insert_with(func);
    |                                           +
-help: consider using a semicolon here
-   |
-LL |         };
-   |          +
 help: you might have meant to return this value
    |
 LL |             return value.get_or_insert_with(func);

--- a/tests/ui/suggestions/try-operator-dont-suggest-semicolon.rs
+++ b/tests/ui/suggestions/try-operator-dont-suggest-semicolon.rs
@@ -18,7 +18,6 @@ fn main() -> Result<(), ()> {
         //~| NOTE: expected `()`, found integer
         //~| HELP: consider using a semicolon here
     }
-    //~^ HELP: consider using a semicolon here
 
     Ok(())
 }

--- a/tests/ui/suggestions/try-operator-dont-suggest-semicolon.stderr
+++ b/tests/ui/suggestions/try-operator-dont-suggest-semicolon.stderr
@@ -12,19 +12,12 @@ error[E0308]: mismatched types
 LL | /     if true {
 LL | |
 LL | |         x?
-   | |         ^^ expected `()`, found integer
+   | |         ^^- help: consider using a semicolon here: `;`
+   | |         |
+   | |         expected `()`, found integer
 ...  |
 LL | |     }
    | |_____- `if` expressions without `else` arms expect their inner expression to be `()`
-   |
-help: consider using a semicolon here
-   |
-LL |         x?;
-   |           +
-help: consider using a semicolon here
-   |
-LL |     };
-   |      +
 
 error: aborting due to 2 previous errors
 

--- a/tests/ui/typeck/issue-112007-leaked-writeln-macro-internals.stderr
+++ b/tests/ui/typeck/issue-112007-leaked-writeln-macro-internals.stderr
@@ -12,10 +12,6 @@ LL | |     }
    |
    = note: expected unit type `()`
                    found enum `Result<(), std::fmt::Error>`
-help: consider using a semicolon here
-   |
-LL |     };
-   |      +
 help: you might have meant to return this value
    |
 LL |         return writeln!(w, "but not here");
@@ -42,10 +38,6 @@ LL | |     }
    = note: expected unit type `()`
                    found enum `Result<(), std::fmt::Error>`
    = note: this error originates in the macro `writeln` which comes from the expansion of the macro `baz` (in Nightly builds, run with -Z macro-backtrace for more info)
-help: consider using a semicolon here
-   |
-LL |     };
-   |      +
 help: you might have meant to return this value
    |
 LL |         return baz!(w);

--- a/tests/ui/typeck/question-mark-operator-suggestion-span.stderr
+++ b/tests/ui/typeck/question-mark-operator-suggestion-span.stderr
@@ -12,10 +12,6 @@ LL | |     }
    |
    = note: expected unit type `()`
                    found enum `Result<(), std::fmt::Error>`
-help: consider using a semicolon here
-   |
-LL |     };
-   |      +
 help: you might have meant to return this value
    |
 LL |         return writeln!(w, "but not here");

